Port cubeb_pulse.c to rust

djg opened the pull request:
This is a direct port of cubeb_pulse.c with light conversion to be "rust-like"

@kinetiknz please review

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+        let r = pa_stream_get_latency(self.output_stream, &mut r_usec, &mut negative);
+
+        debug_assert!(negative == 0);
+        if r != 0 {
```
We shouldn't really assert the value of negative until after we've checked the return code, since the callee could write whatever it likes in there (I know this is wrong in the original code also)

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+    }
+
+    ///
+    ///
```
Remove these?

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+            debug_assert!(size > 0);
+            debug_assert!(size % frame_size == 0);
+
+            //LOGV("Trigger user callback with output buffer size=%zd, read_offset=%zd", size, read_offset);
```
Re-enable this logging?

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+
+unsafe extern fn stream_success_callback(_s: *mut pa_stream, _success: i32, u: *mut c_void)
+{
+  let stm = &*(u as *mut Stream);
```
Indentation is much shallower for this and some of the following code vs the elsewhere

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+    if stm.input_stream.is_null() {
+        // Output/playback only operation.
+        // Write directly to output
+        debug_assert!(!!stm.input_stream.is_null() && !stm.output_stream.is_null());
```
is_null returns bool, so don't need the double-inversion here

djg replied:
Checking input_stream.is_null feels a bit suspicious since the `if stm.input_stream.is_null() {`

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+
+unsafe extern fn stream_read_callback(s: *mut pa_stream, _nbytes: usize, u: *mut c_void)
+{
+    //  LOGV("Input callback buffer size %zd", nbytes);
```
(For all LOGVs) can we re-enable these?

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+                stm.trigger_user_callback(stream, read_data, write_size);
+            } else {
+                // input/capture only operation. Call callback directly
+                let got = stm.data_callback.unwrap()(stm as *const _ as *mut _,
```
Why -> const -> mut?

djg replied:
You're right. It should be stm as *mut _ as * mut _. First to convert &mut Stream to *mut Stream and second to convert *mut Stream to *mut c_void.

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+    }
+}
+
+//
```
Remove?

kinetiknz commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+    true
+}
+
+//
```
Remove?

djg commented on cubeb-ffi/src/ffi.rs:
```diff
@@ -0,0 +1,674 @@
+            }
+        }
+    };
+    ($level: expr, $fmt: expr, $($arg:tt)+) => {
```
I'll run both tools.

djg commented on cubeb-ffi/src/ffi.rs:
```diff
@@ -0,0 +1,674 @@
+
+// TODO endian check
+pub const SAMPLE_S16NE: SampleFormat = SampleFormat::S16LE;
+pub const SAMPLE_FLOAT32NE: SampleFormat = SampleFormat::Float32LE;
```
It appears these values are used in cubeb_jack.cpp, cubeb_resampler.cpp and test_resampler.cpp in cubeb. I had planned to share cubeb-ffi between different back ends (if we go that far).

Do you still suggest removing these?

kinetiknz replied:
Ah, it wasn't clear these were part of cubeb_sample_format since the other members are part of enum SampleFormat.  It's a bit weird this way since they have such different naming, but I guess it's okay for now.

djg commented on Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+name = "cubeb-pulse"
+version = "0.0.1"
+authors = ["Dan Glastonbury <dglastonbury@mozilla.com>"]
+description = "Cubeb backed for Pulse Audio written in Rust"
```
My plan was that each backend would live in it's own crate. That seems to be the rust way. (Many, small crates).

djg commented on cubeb-ffi/src/ffi.rs:
```diff
@@ -0,0 +1,674 @@
+#[macro_export]
+macro_rules! log_internal {
+    ($level: expr, $msg: expr) => {
+        #[allow(unused_unsafe)]
```
Yes. If you place log! or logv! in an unsafe block then you'll get a warning that the unsafe is not needed.

djg commented:
I've cleaned this up now. The issue I had was that I couldn't get #[cfg()]
to work inside an enum.

The change is convert enum SampleFormat {} into struct SampleFormat(i32)
and make each enum value into const SAMPLE_S16: SampleFormat =
SampleFormat(0); etc.

This matches the cubeb C-style much more. I dropped the CUBEB_ because I
prefer using the module name as a prefix, ie. cubeb::SAMPLE_S16LE instead
of CUBEB_SAMPLE_S16LE.

On 31 Mar. 2017 11:27, "Matthew Gregan" <notifications@github.com> wrote:

> *@kinetiknz* commented on this pull request.
> ------------------------------
>
> In cubeb-ffi/src/ffi.rs
> <https://github.com/djg/cubeb-pulse-rs/pull/1#discussion_r109072623>:
>
> > +}
> +
> +#[macro_export]
> +macro_rules! log {
> +    ($msg: expr) => (log_internal!($crate::LogLevel::Normal, $msg));
> +    ($fmt: expr, $($arg: tt)+) => (log_internal!($crate::LogLevel::Normal, $fmt, $($arg)*));
> +}
> +
> +//     g_log_callback("%s:%d: " fmt "\n", __FILE__, __LINE__, ##__VA_ARGS__); \
> +
> +pub enum Context {}
> +pub enum Stream {}
> +
> +// TODO endian check
> +pub const SAMPLE_S16NE: SampleFormat = SampleFormat::S16LE;
> +pub const SAMPLE_FLOAT32NE: SampleFormat = SampleFormat::Float32LE;
>
> Ah, it wasn't clear these were part of cubeb_sample_format since the other
> members are part of enum SampleFormat. It's a bit weird this way since they
> have such different naming, but I guess it's okay for now.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/djg/cubeb-pulse-rs/pull/1#discussion_r109072623>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAJhrTfOhNsO_MOdtfNZSEuOAxx2R0hWks5rrFZ8gaJpZM4Me1eK>
> .
>

djg commented on cubeb-ffi/src/ffi.rs:
```diff
@@ -0,0 +1,674 @@
+
+extern "C" {
+    pub static g_log_level: LogLevel;
+    pub static g_log_callback: LogCallback;
```
These aren't exported in the public interface of cubeb.h. I'll open an issue against cubeb.

djg commented on src/capi.rs:
```diff
@@ -0,0 +1,212 @@
+
+extern "C" fn capi_init(c: *mut *mut cubeb::Context, context_name: *const i8) -> i32
+{
+    unsafe {
```
I think that's a fair assessment. I've added a lot of unsafe blocks around the PA calls in the backend, but I want to leave those for now, as I've explained, I intend to replace the raw FFI calls with more ergonomic rust wrappers.

djg commented on src/backend/context.rs:
```diff
@@ -0,0 +1,706 @@
+use std::ffi::CStr;
+use std::ptr;
+
+macro_rules! dup_str {
```
Did you peak ahead at the next chapter? ;-)

djg commented on src/backend/context.rs:
```diff
@@ -0,0 +1,706 @@
+    }
+}
+
+const PA_RATE_MAX: u32 = 48000*8;
```
I think bindgen has skipped some #defines from pulse. I'll move it.

djg commented on src/backend/context.rs:
```diff
@@ -0,0 +1,706 @@
+        pa_threaded_mainloop_start(ctx.mainloop);
+
+        if ctx.pulse_context_init() != cubeb::OK {
+            ctx.destroy();
```
It probably is, but to aid the review, I didn't want to change too much of the structure of the C->rust code in this pass.

djg commented on src/backend/context.rs:
```diff
@@ -0,0 +1,706 @@
+
+        let o = pa_context_get_server_info(self.context,
+                                           Some(pulse_server_info_cb),
+                                           &mut user_data as *mut _ as *mut _);
```
Yeah, unfortunately they do.

`&mut user_data as *mut _` results in type `*mut PulseDevListData`, the second `as *mut _` results in type `*mut c_void`. You can't jump from `&mut t` to `*mut c_void` in one step in rust, otherwise rustc will report error:

casting `&mut backend::context::PulseDevListData` as `*mut std::os::raw::c_void` is invalid

djg commented on src/backend/context.rs:
```diff
@@ -0,0 +1,706 @@
+        true
+    }
+
+    pub fn wait_until_context_ready(&self) -> bool
```
I've opened an issue to look at this later.

djg commented on src/backend/var_array.rs:
```diff
@@ -0,0 +1,55 @@
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct VarArray<T>
```
This was the trickiest and most unsafe part of the port. I'd love to do with out it.

djg commented on src/backend/var_array.rs:
```diff
@@ -0,0 +1,55 @@
+            Some(box_ptr) => Box::into_raw(box_ptr) as *mut u8,
+            None => 0 as *mut u8
+        };
+        let mem = pa_xrealloc(raw_ptr as *mut _, size);
```
Because the C version has access to malloc/free, etc. It was convenient to use PA's zeroing wrappers for lib realloc, et al., because those are exposed, especially after removing the dependency upon libc.

djg commented on src/backend/stream.rs:
```diff
@@ -0,0 +1,821 @@
+
+const PULSE_NO_GAIN: f32 = -1.0;
+const PA_USEC_PER_MSEC: pa_usec_t = 1000;
+const PA_USEC_PER_SEC: pa_usec_t = 1000000;
```
I remember why I did this initially. I was using the libpulse-sys crate that was missing these values. Now that has been replaced, I'll add the definitions to pulse-ffi crate.

djg commented on cubeb-ffi/src/ffi.rs:
```diff
@@ -0,0 +1,674 @@
+}
+
+#[repr(C)]
+#[derive(Clone,Copy,Debug)]
```
Rustfmt doesn't appear to handle this case. 👎

djg commented on LICENSE:
```diff
@@ -0,0 +1,13 @@
+Copyright © 2011 Mozilla Foundation
```
I updated all the source code headers to contain ISC license instead of MPL 2.0 header.
